src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Build/Diagnostics/IncrementalBuildFailureTelemetryReporter.cs

```diff
@@ -35,9 +33,7 @@ public async Task<bool> IsEnabledAsync(CancellationToken cancellationToken)
                 return false;
             }
 
-            await _project.Services.ThreadingPolicy.SwitchToUIThread(cancellationToken);
-
-            IVsFeatureFlags featureFlagsService = _featureFlagsService.Value;
+            IVsFeatureFlags featureFlagsService = await _featureFlagsService.GetValueAsync(cancellationToken);
```

Comment:
Unfortunately, this doesn't completely resolve the problem; Andrew's versions of IVsService aren't cached between importers; everyone gets a unique version. Currently every GetServiceAsync does a switch to the UI thread [for the cast](http://index/?leftProject=Microsoft.VisualStudio.Utilities&leftSymbol=j4dykj4yoqj1&rightProject=Microsoft.VisualStudio.Shell.15.0&file=ServiceExtensions.cs&line=144), so you still pay for the UI thread once per project for this object.

That might not sound too bad, but we're actually seeing this add up in the wild in https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1827681/?view=edit for example.

Reply:
I would have this call through IProjectSystemOptions. You haven't made this worse so don't block on it but something to remember for the future.

Reply:
530b2cf772697cc9971a4e80a14333125c8cec44 consolidates all access to `IVsService<IVsFeatureFlags>` within `ProjectSystemOptions` and uses `ValueTask<bool>` for the accessor methods so we shouldn't see any allocation or thread switching beyond the very first call in the IDE (that component lives in ProjectService scope.)